### PR TITLE
added: ability to open the smartplaylisteditor for a given playlist

### PR DESCRIPTION
**Pull request by notspiff**
Requested in forums, ref http://forum.kodi.tv/showthread.php?tid=224034

### CODE REVIEW

**Comment by Montellese** [comment]
Maybe I'm missing something but why are you guessing for the type? If the XSP file exists it should also contain the type of the smartplaylist.

**Comment by akva2** [comment]
i'm guessing based on the folder names.

special://videoplaylists/blah.xsp
special://musicplaylists/blah.xsp

sure, we could open the xsp instead in principle. in any case the editor requires that we specify this up front for which controls are visible etc. not my invention.

**Comment by Montellese** [comment]
Yes but for that you need to set `m_mode` which I don't see happening. And you load the playlist and then set the type even though you can already get it from the loaded playlist at that point.

**Comment by notspiff** [comment]
I dont do any of those things.

**Comment by notspiff** [comment]
As in read again and consider *nonexistent* files.

File: xbmc/dialogs/GUIDialogSmartPlaylistEditor.cpp [hunk 1]
```diff
@@ -142,6 +143,31 @@ bool CGUIDialogSmartPlaylistEditor::OnMessage(CGUIMessage& message)
...
+  case GUI_MSG_WINDOW_INIT:
+    {
+      const std::string& startupList = message.GetStringParam(1);
+      std::string type = message.GetStringParam(2);
+      if (!startupList.empty())
+      {
+        // guess type
+        if (type.empty())
+        {
+          if (startupList.find("video") > -1)
```
Comment (Montellese) [inline xbmc/dialogs/GUIDialogSmartPlaylistEditor.cpp]: Shouldn't this be `!= std::string::npos` (same below)?

Comment (notspiff) [inline xbmc/dialogs/GUIDialogSmartPlaylistEditor.cpp]: indeed. used to be cstdstring code..

**Comment by Montellese** [comment]
OK for non-existing files that makes sense. But it still looks to me like you need to
* set `m_mode` (which in case of an existing smartplaylist and no type parameter could be determined from the type of the loaded smartplaylist)
* only change the playlist type if `m_playlist.Load()` fails

**Comment by notspiff** [comment]
rebased and fixed. can't get it less ugly.

**Comment by notspiff** [comment]
note that i removed the ability to open new files due to API issues (only a single string parameter is passed on from builtins/event server).

File: xbmc/dialogs/GUIDialogSmartPlaylistEditor.cpp [hunk 2]
```diff
@@ -142,6 +143,39 @@ bool CGUIDialogSmartPlaylistEditor::OnMessage(CGUIMessage& message)
...
+      const std::string& startupList = message.GetStringParam(0);
+      if (!startupList.empty())
+      {
+        int party = 0;
+        if (URIUtils::PathEquals(startupList, CProfilesManager::Get().GetUserDataItem("PartyMode.xsp")))
+          party = 1;
+        if (URIUtils::PathEquals(startupList, CProfilesManager::Get().GetUserDataItem("PartyMode-Video.xsp")))
+          party = 2;
+
+        if (party || m_playlist.Load(startupList))
```
Comment (Montellese) [inline xbmc/dialogs/GUIDialogSmartPlaylistEditor.cpp]: You still need to load the playlist if it is a partymode playlist. I'd probably go with resetting `m_mode` at the beginning, then setting it to either `partymusic` or `partyvideo` if one of those conditions matches and then after having loaded the playlist, set `m_mode` to `music` or `video` (using the conditions you have in place) if it hasn't been set before.

Does that make sense?

Comment (notspiff) [inline xbmc/dialogs/GUIDialogSmartPlaylistEditor.cpp]: fixed otherwise.

**Comment by Montellese** [comment]
Should be good now, thanks.

**Comment by Lunatixz** [comment]
Thanks for adding this feature.

**Comment by Lunatixz** [comment]
So will this make it into master in some form?

**Comment by NedScott** [comment]
Most likely, yes. Spiff doesn't want to handle the PRs himself, so he went and closed them, but we're still keeping track of them. We just have to wait for master to be un-feature frozen.

**Comment by Lunatixz** [comment]
Thanks :+1:

**Pull request closed without merging**